File: triathlon-trainer-web/src/services/weatherService.ts
import type { HourlyForecast, WeatherAssessment, WeatherWarning, RiskLevel, Sport } from '../types'

const API_KEY = import.meta.env.VITE_OPENWEATHER_API_KEY || ''
const BASE = 'https://api.openweathermap.org/data/3.0'

// ─── Fetch ────────────────────────────────────────────────────────────────

export async function fetchForecast(lat: number, lon: number): Promise<HourlyForecast[]> {
  if (!API_KEY) return []

  const res = await fetch(
    `${BASE}/onecall?lat=${lat}&lon=${lon}&exclude=minutely,daily,alerts&units=metric&appid=${API_KEY}`
  )
  if (!res.ok) return []

  const data = await res.json()
  return (data.hourly ?? []).slice(0, 48).map((h: any): HourlyForecast => ({
    dt: h.dt * 1000,
    temp: h.temp,
    feelsLike: h.feels_like,
    description: h.weather?.[0]?.description ?? '',
    windKmh: (h.wind_speed ?? 0) * 3.6,
    precipChance: h.pop ?? 0,
    precipMm: (h.rain?.['1h'] ?? h.snow?.['1h'] ?? 0),
    icon: h.weather?.[0]?.icon ?? '',
  }))
}

export async function getCurrentLocation(): Promise<{ lat: number; lon: number } | null> {
  return new Promise((resolve) => {
    if (!navigator.geolocation) { resolve(null); return }
    navigator.geolocation.getCurrentPosition(
      (pos) => resolve({ lat: pos.coords.latitude, lon: pos.coords.longitude }),
      () => resolve(null),
      { timeout: 8000 }
    )
  })
}

// ─── Assessment ───────────────────────────────────────────────────────────

export function assessWeather(
  forecast: HourlyForecast[],
  sport: Sport,
  workoutDate: string
): WeatherAssessment {
  const target = new Date(workoutDate).getTime()
  const conditions = forecast.reduce((best, h) =>
    Math.abs(h.dt - target) < Math.abs(best.dt - target) ? h : best,
    forecast[0]
  )

  if (!conditions) {
    return { isSuitable: true, riskLevel: 'unknown', warnings: [], indoorRecommended: false, summary: 'Weather data unavailable — check before heading out.' }
  }

  const warnings: WeatherWarning[] = []
  let risk: RiskLevel = 'low'

  const maxRisk = (r: RiskLevel) => {
    const order: RiskLevel[] = ['unknown', 'low', 'moderate', 'high', 'extreme']
    if (order.indexOf(r) > order.indexOf(risk)) risk = r
  }

  // Temperature
  if (conditions.feelsLike < -10) {
    warnings.push({ icon: '🥶', message: `Extreme cold (${Math.round(conditions.feelsLike)}°C). High hypothermia risk.` })
    maxRisk('extreme')
  } else if (conditions.feelsLike < 0) {
    warnings.push({ icon: '❄️', message: `Very cold (${Math.round(conditions.feelsLike)}°C). Dress in layers, protect extremities.` })
    maxRisk('high')
  } else if (conditions.feelsLike > 35) {
    warnings.push({ icon: '🌡️', message: `Extreme heat (${Math.round(conditions.feelsLike)}°C). High heat stroke risk.` })
    maxRisk('extreme')
  } else if (conditions.feelsLike > 30) {
    warnings.push({ icon: '☀️', message: `Very hot (${Math.round(conditions.feelsLike)}°C). Reduce intensity, carry extra fluids.` })
    maxRisk('high')
  } else if (conditions.feelsLike > 27) {
    warnings.push({ icon: '🌤️', message: `Hot (${Math.round(conditions.feelsLike)}°C). Slow down 10–15%, double hydration.` })
    maxRisk('moderate')
  }

  // Wind
  if (conditions.windKmh > 60) {
    warnings.push({ icon: '💨', message: `Dangerous wind (${Math.round(conditions.windKmh)} km/h). Debris risk — stay indoors.` })
    maxRisk('extreme')
  } else if (conditions.windKmh > 40) {
    warnings.push({ icon: '💨', message: `Strong wind (${Math.round(conditions.windKmh)} km/h). Outdoor cycling not recommended.` })
    if (sport === 'Bike') maxRisk('high')
    else maxRisk('moderate')
  } else if (conditions.windKmh > 25 && sport === 'Bike') {
    warnings.push({ icon: '🌬️', message: `Gusty (${Math.round(conditions.windKmh)} km/h). Plan route with wind at back on return.` })
  }

  // Rain
  if (conditions.precipChance > 0.7) {
    warnings.push({ icon: '🌧️', message: `High rain chance (${Math.round(conditions.precipChance * 100)}%). Wet roads — reduce bike speed in corners.` })
    if (sport === 'Bike') maxRisk('moderate')
  }
  if (conditions.precipMm > 10) {
    warnings.push({ icon: '⛈️', message: `Heavy rain (${Math.round(conditions.precipMm)}mm). Roads may flood.` })
    maxRisk('high')
  }

  // Storm
  if (conditions.description.toLowerCase().includes('thunder') || conditions.description.toLowerCase().includes('storm')) {
    warnings.push({ icon: '⚡', message: 'Thunderstorm forecast. No open water swimming. Avoid high ground on bike.' })
    maxRisk('extreme')
  }

  const finalRisk = risk as RiskLevel
  const indoorRecommended = finalRisk === 'extreme' || finalRisk === 'high'
  const intensityAdjustment = conditions.feelsLike > 27 ? 'Reduce target pace/power by 10–15% for the heat' : undefined

  const summary = {
    unknown: 'Weather data unavailable.',
    low: `Good conditions for your ${sport.toLowerCase()}.`,
    moderate: 'Manageable — take precautions and adjust effort as needed.',
    high: 'Challenging conditions. Consider the indoor alternative or reschedule.',
    extreme: 'Dangerous conditions. Indoor training strongly recommended.',
  }[finalRisk]

  return { isSuitable: finalRisk !== 'extreme', riskLevel: finalRisk, warnings, indoorRecommended, intensityAdjustment, summary }
}

// ─── Indoor Alternatives ──────────────────────────────────────────────────

export function getIndoorAlternative(sport: Sport) {
  switch (sport) {
    case 'Bike':
    case 'Brick':
      return {
        instructions: 'Complete on your indoor trainer (turbo / smart trainer). Same duration and effort zones as planned. Use structured ERG mode if available.\n\nFor Zwift/TrainerRoad: load a workout matching today\'s intensity zones.\nFor unstructured riding: use perceived effort, maintain cadence 85–95 rpm.',
        equipment: ['Indoor trainer / turbo', 'Fan (mandatory — no wind indoors)', 'Towel', '2× water bottles', 'Chamois cream'],
      }
    case 'Run':
      return {
        instructions: 'Treadmill alternative — same session indoors. Set treadmill incline to 0.5–1% to simulate outdoor air resistance.\n\nFor intervals: pre-set your interval speed and step on/off the belt for rest intervals.',
        equipment: ['Treadmill'],
      }
    default:
      return {
        instructions: 'Home bodyweight strength and core session. 3×15 of: plank, glute bridge, single-leg deadlift, press-up, mountain climber.',
        equipment: ['Exercise mat'],
      }
  }
}

File: TriathlonTrainer/Sources/TriathlonTrainer/Services/WeatherService.swift
import Foundation
import WeatherKit
import CoreLocation

/// Fetches weather forecasts via Apple WeatherKit and evaluates outdoor training suitability.
@MainActor
class WeatherService: ObservableObject {

    private let weatherService = WeatherService.shared
    // Note: WeatherKit.WeatherService is used above — alias to avoid name conflict
    private let wkService = WeatherKit.WeatherService()

    @Published var currentConditions: CurrentConditions?
    @Published var hourlyForecast: [HourlyConditions] = []
    @Published var isLoading = false
    @Published var error: String?

    // MARK: - Fetch Forecast

    func fetchForecast(for location: CLLocation) async {
        isLoading = true
        error = nil
        defer { isLoading = false }

        do {
            let weather = try await wkService.weather(for: location)

            currentConditions = CurrentConditions(
                temperature: weather.currentWeather.temperature.converted(to: .celsius).value,
                feelsLike: weather.currentWeather.apparentTemperature.converted(to: .celsius).value,
                condition: weather.currentWeather.condition.description,
                windSpeedKmh: weather.currentWeather.wind.speed.converted(to: .kilometersPerHour).value,
                humidity: weather.currentWeather.humidity,
                uvIndex: Int(weather.currentWeather.uvIndex.value),
                isDaylight: weather.currentWeather.isDaylight
            )

            hourlyForecast = weather.hourlyForecast.forecast.prefix(48).map { hour in
                HourlyConditions(
                    date: hour.date,
                    temperature: hour.temperature.converted(to: .celsius).value,
                    feelsLike: hour.apparentTemperature.converted(to: .celsius).value,
                    condition: hour.condition.description,
                    windSpeedKmh: hour.wind.speed.converted(to: .kilometersPerHour).value,
                    precipitationChance: hour.precipitationChance,
                    precipitationAmount: hour.precipitationAmount.converted(to: .millimeters).value
                )
            }
        } catch {
            self.error = "Unable to fetch weather: \(error.localizedDescription)"
        }
    }

    // MARK: - Workout Weather Assessment

    /// Evaluates whether conditions are suitable for an outdoor workout of a given sport.
    func assess(sport: Sport, on date: Date) -> WeatherAssessment {
        guard let conditions = hourlyConditions(for: date) else {
            return WeatherAssessment(
                isSuitable: true,
                riskLevel: .unknown,
                warnings: [],
                indoorRecommended: false,
                adjustedIntensity: nil,
                summary: "Weather data unavailable — check before heading out."
            )
        }

        var warnings: [WeatherWarning] = []
        var riskLevel: RiskLevel = .low

        // --- Temperature checks ---
        switch conditions.feelsLike {
        case ..<(-10):
            warnings.append(WeatherWarning(icon: "thermometer.snowflake", message: "Extreme cold (\(Int(conditions.feelsLike))°C). High hypothermia risk outdoors."))
            riskLevel = .extreme
        case -10..<0:
            warnings.append(WeatherWarning(icon: "thermometer.snowflake", message: "Very cold (\(Int(conditions.feelsLike))°C). Dress in layers, protect extremities."))
            riskLevel = max(riskLevel, .high)
        case 35...:
            warnings.append(WeatherWarning(icon: "thermometer.sun", message: "Extreme heat (\(Int(conditions.feelsLike))°C). High heat stroke risk."))
            riskLevel = .extreme
        case 30..<35:
            warnings.append(WeatherWarning(icon: "thermometer.sun", message: "Very hot (\(Int(conditions.feelsLike))°C). Reduce intensity, carry extra fluids."))
            riskLevel = max(riskLevel, .high)
        case 27..<30:
            warnings.append(WeatherWarning(icon: "thermometer", message: "Hot conditions (\(Int(conditions.feelsLike))°C). Slow down 10-15%, double hydration."))
            riskLevel = max(riskLevel, .moderate)
        default:
            break
        }

        // --- Wind checks ---
        if conditions.windSpeedKmh > 60 {
            warnings.append(WeatherWarning(icon: "wind", message: "Dangerous wind (\(Int(conditions.windSpeedKmh)) km/h). Risk of being knocked over or debris."))
            riskLevel = max(riskLevel, .extreme)
        } else if conditions.windSpeedKmh > 40 {
            warnings.append(WeatherWarning(icon: "wind", message: "Strong wind (\(Int(conditions.windSpeedKmh)) km/h). Bike outdoors not recommended."))
            if sport == .bike { riskLevel = max(riskLevel, .high) }
            else { riskLevel = max(riskLevel, .moderate) }
        } else if conditions.windSpeedKmh > 25 && sport == .bike {
            warnings.append(WeatherWarning(icon: "wind", message: "Gusty (\(Int(conditions.windSpeedKmh)) km/h). Plan route to have wind at back on return leg."))
            riskLevel = max(riskLevel, .low)
        }

        // --- Precipitation checks ---
        if conditions.precipitationChance > 0.7 {
            let icon = sport == .bike ? "cloud.rain" : "cloud.drizzle"
            warnings.append(WeatherWarning(icon: icon, message: "High rain chance (\(Int(conditions.precipitationChance * 100))%). Wet roads — reduce bike speed, especially corners."))
            if sport == .bike { riskLevel = max(riskLevel, .moderate) }
        }

        if conditions.precipitationAmount > 10 {
            warnings.append(WeatherWarning(icon: "cloud.heavyrain", message: "Heavy rain forecast (\(Int(conditions.precipitationAmount))mm). Roads may flood, visibility poor."))
            riskLevel = max(riskLevel, .high)
        }

        // --- Lightning/Thunder ---
        if conditions.condition.lowercased().contains("thunder") || conditions.condition.lowercased().contains("storm") {
            warnings.append(WeatherWarning(icon: "bolt.fill", message: "Thunderstorm in forecast. Do not swim in open water, avoid high ground on bike."))
            riskLevel = .extreme
        }

        // --- UV checks ---
        if let current = currentConditions, current.uvIndex >= 8 {
            warnings.append(WeatherWarning(icon: "sun.max", message: "Very high UV (index \(current.uvIndex)). Apply SPF 50+, wear a cap, consider morning training instead."))
            riskLevel = max(riskLevel, .moderate)
        }

        // --- Build assessment ---
        let indoorRecommended = riskLevel == .extreme || riskLevel == .high
        let adjustedIntensity: String? = conditions.feelsLike > 27 ? "Reduce target pace/power by 10-15% for heat" : nil

        let summary: String
        switch riskLevel {
        case .low, .unknown:
            summary = "Conditions look good for your \(sport.rawValue.lowercased()) session."
        case .moderate:
            summary = "Manageable conditions — take precautions and adjust effort as needed."
        case .high:
            summary = "Challenging conditions. Consider the indoor alternative or reschedule."
        case .extreme:
            summary = "Dangerous conditions. Indoor training strongly recommended today."
        }

        return WeatherAssessment(
            isSuitable: riskLevel != .extreme,
            riskLevel: riskLevel,
            warnings: warnings,
            indoorRecommended: indoorRecommended,
            adjustedIntensity: adjustedIntensity,
            summary: summary
        )
    }

    // MARK: - Helpers

    func hourlyConditions(for date: Date) -> HourlyConditions? {
        hourlyForecast.min(by: { abs($0.date.timeIntervalSince(date)) < abs($1.date.timeIntervalSince(date)) })
    }

    func snapshotForWorkout(sport: Sport, on date: Date) -> WeatherSnapshot {
        guard let conditions = hourlyConditions(for: date) else {
            return WeatherSnapshot(
                temperature: 20, feelsLike: 20, condition: "Unknown",
                windSpeedKmh: 0, precipitationProbability: 0, isOutdoorSafe: true
            )
        }

        let assessment = assess(sport: sport, on: date)

        return WeatherSnapshot(
            temperature: conditions.temperature,
            feelsLike: conditions.feelsLike,
            condition: conditions.condition,
            windSpeedKmh: conditions.windSpeedKmh,
            precipitationProbability: conditions.precipitationChance,
            isOutdoorSafe: assessment.isSuitable,
            warningMessage: assessment.warnings.first?.message
        )
    }
}

// MARK: - Supporting Types

struct CurrentConditions {
    var temperature: Double
    var feelsLike: Double
    var condition: String
    var windSpeedKmh: Double
    var humidity: Double
    var uvIndex: Int
    var isDaylight: Bool
}

struct HourlyConditions {
    var date: Date
    var temperature: Double
    var feelsLike: Double
    var condition: String
    var windSpeedKmh: Double
    var precipitationChance: Double
    var precipitationAmount: Double
}

struct WeatherAssessment {
    var isSuitable: Bool
    var riskLevel: RiskLevel
    var warnings: [WeatherWarning]
    var indoorRecommended: Bool
    var adjustedIntensity: String?
    var summary: String
}

struct WeatherWarning {
    var icon: String
    var message: String
}

enum RiskLevel: Int, Comparable {
    case unknown = 0, low = 1, moderate = 2, high = 3, extreme = 4

    static func < (lhs: RiskLevel, rhs: RiskLevel) -> Bool {
        lhs.rawValue < rhs.rawValue
    }

    static func max(_ lhs: RiskLevel, _ rhs: RiskLevel) -> RiskLevel {
        lhs > rhs ? lhs : rhs
    }
}

// MARK: - Indoor Alternatives

extension WeatherService {
    /// Returns an indoor alternative workout for when conditions are poor.
    static func indoorAlternative(for workout: Workout) -> IndoorAlternative {
        switch workout.sport {
        case .bike, .brick:
            return IndoorAlternative(
                workoutType: .bikeIndoorTrainer,
                instructions: """
                    Complete this session on your indoor trainer (turbo trainer / smart trainer). \
                    Same duration and effort zones as planned. Use structured ERG mode if available. \
                    \n\nFor Zwift/TrainerRoad users: load a workout matching today's intensity zones. \
                    \nFor unstructured riding: use perceived effort and maintain cadence 85-95 rpm.
                    """,
                equipmentNeeded: ["Indoor trainer / turbo", "Fan (mandatory — no wind indoors)", "Towel", "Water bottles x2", "Chamois cream"]
            )
        case .run:
            return IndoorAlternative(
                workoutType: .runEasy,
                instructions: """
                    Treadmill alternative — same session indoors. \
                    Set treadmill incline to 0.5-1% to simulate outdoor air resistance. \
                    \n\nFor intervals: treadmill speed is slower to change than outdoor pace — \
                    pre-set your interval speed and jump on/off the belt for rest intervals.
                    """,
                equipmentNeeded: ["Treadmill"]
            )
        case .swim:
            return IndoorAlternative(
                workoutType: .swimTechnique,
                instructions: "Pool swimming is always indoor — proceed with the planned session.",
                equipmentNeeded: ["Pool access", "Goggles", "Swim cap"]
            )
        default:
            return IndoorAlternative(
                workoutType: .strengthCore,
                instructions: "Home bodyweight strength and core session instead. 3×15 of: plank, glute bridge, single-leg deadlift, press-up, mountain climber.",
                equipmentNeeded: ["Exercise mat"]
            )
        }
    }
}
